drivers/at_cmd/at_cmd.c

```diff
@@ -0,0 +1,288 @@
+static at_cmd_handler_t received_data_handler;
+
+struct callback_work_item {
+	struct k_work    work;
+	u32_t            data_len;
+	char             data[RESPONSE_MAX_LEN];
+	at_cmd_handler_t callback;
+};
+
+static int open_socket(void)
+{
+	common_socket_fd = socket(AF_LTE, 0, NPROTO_AT);
+
+	if (common_socket_fd == -1) {
+		LOG_ERR("AT socket creation failed");
+		return -EFAULT;
+	}
+
+	return 0;
+}
+
+static enum at_cmd_rc get_return_code(char *buf)
+{
+	enum at_cmd_rc return_code = AT_CMD_NO_RETURN_CODE;
+	char *tmpstr = NULL;
+
+	do {
+		tmpstr = strstr(buf, AT_CMD_OK_STR);
+		if (tmpstr) {
+			return_code = AT_CMD_OK;
+			break;
+		}
+
+		tmpstr = strstr(buf, AT_CMD_ERROR_STR);
+		if (tmpstr) {
+			return_code = AT_CMD_ERROR;
+			break;
+		}
+
+		tmpstr = strstr(buf, AT_CMD_CMS_STR);
+		if (tmpstr) {
+			return_code = AT_CMD_CMS;
+			break;
+		}
+
+		tmpstr = strstr(buf, AT_CMD_CME_STR);
+		if (tmpstr) {
+			return_code = AT_CMD_CME;
+			break;
+		}
+	} while (0);
+
+	if ((return_code == AT_CMD_CMS) || (return_code == AT_CMD_CME)) {
+		last_error = atoi(&buf[ARRAY_SIZE(AT_CMD_CMS_STR) - 1]);
+	} else {
+		last_error = 0;
+	}
+
+	if (tmpstr) {
+		buf[tmpstr - buf] = '\0';
+	}
+
+	return return_code;
+}
+
+static void callback_worker(struct k_work *item)
+{
+	struct callback_work_item *data =
+		CONTAINER_OF(item, struct callback_work_item, work);
+
+	if ((data != NULL) && (data->callback != NULL)) {
+		data->callback(data->data, data->data_len);
+	}
+
+	k_free(data);
+}
+
+
+static void socket_thread_fn(void *arg1, void *arg2, void *arg3)
+{
+	int bytes_read;
+	char buf[RESPONSE_MAX_LEN];
+	enum at_cmd_rc return_code;
+	bool write_with_response = false;
+	at_cmd_handler_t current_callback = NULL;
+
+	ARG_UNUSED(arg1);
+	ARG_UNUSED(arg2);
+	ARG_UNUSED(arg3);
+
+	LOG_DBG("Socket thread started");
+
+	for (;;) {
+		bytes_read = recv(common_socket_fd, buf, sizeof(buf), 0);
+
+		if (bytes_read < 0) {
+			LOG_ERR("Socket recv failed with err %d", bytes_read);
+
+			if (bytes_read == -EAGAIN) {
+				if (open_socket() == 0) {
+					goto failed;
+				}
+			}
+
+			LOG_ERR("Unrecoverable reception error, thread killed");
+			return;
+		} else if (bytes_read == sizeof(buf) &&
+			   buf[bytes_read - 1] != '\0') {
+
+			LOG_DBG("AT message to large for reception buffer");
+			goto failed;
+		}
+
+		return_code = get_return_code(buf);
+		current_callback = received_data_handler;
+
+		if (return_code != AT_CMD_NO_RETURN_CODE) {
+			if (response_buf_len > 0) {
+				if (response_buf_len > strlen(buf)) {
+					strncpy(response_buf,
+						buf,
+						response_buf_len);
+				} else {
+					LOG_ERR("Response buffer not large enough");
+					response_buf = NULL;
+				}
+
+				write_with_response = true;
+				response_buf_len = 0;
+			}
+
+			k_msgq_put(&return_code_msq, &return_code, K_FOREVER);
+		} else {
+			write_with_response = false;
+		}
+
+		if (strlen(buf) == 0) {
+			continue;
+		}
+
+		if (received_data_handler) {
```

Comment:
received_data_handler or current_callback ?

Reply:
The null pointer check is moved inside the worker function instead